### PR TITLE
0J-2347: Create Abandon Page

### PR DESCRIPTION
**Pull request by MarieseMikely**
[OJ-2347]Create abandon page

## Proposed changes
Created the abandon page 

### What changed
Added the abandon page and associated feature tests 

### Why did it change
To be included in sprint work.

### Issue tracking

[OJ-2347](https://govukverify.atlassian.net/browse/OJ-2347)

### Environment variables or secrets

- [x] No environment variables or secrets were added or changed

### Other considerations

- [ ] ~Update [README](./blob/main/README.md) with any new instructions or tasks~

[OJ-2347]: https://govukverify.atlassian.net/browse/OJ-2347?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[OJ-2347]: https://govukverify.atlassian.net/browse/OJ-2347?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by liamStockwellGDS** [comment]
Unit tests are pushed now :slightly_smiling_face: - the last browser test is failing due to https://govukverify.atlassian.net/browse/OJ-2346 which 403 is not returned yet from the /authorization endpoint

File: tests/imposter/private-api.yaml [hunk 1]
```diff
@@ -21,6 +21,8 @@ paths:
             application/json:
               schema:
                 $ref: "#/components/schemas/Error"
+        "403":
```
Comment (merlinc) [inline tests/imposter/private-api.yaml:24]: This needs to be added into the appropriate place in the [API private-api.yml](https://github.com/govuk-one-login/ipv-cri-check-hmrc-api/blob/main/infrastructure/private-api.yaml#L18-L29) file

**Review by merlinc — CHANGES_REQUESTED** [review]
LGTM, just needs the API config to be updated in the check-api repo, and then it can be merged

File: tests/browser/pages/abandon.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  async selectReturn() {
+    await this.page
+      .locator(`input[type="radio"][value="retryNationalInsurance"]`)
+      .first()
+      .check();
+  }
+
+  hasErrorQueryParams() {
```
Comment (Caitlin-cooling) [inline tests/browser/pages/abandon.js:38]: Is this method being used?

Comment (Caitlin-cooling) [reply tests/browser/pages/abandon.js:38]: Same question for `isErrorCode` too actually

File: tests/imposter/private-api-config.yaml [hunk 3]
```diff
@@ -146,3 +190,10 @@ resources:
             "error": "server_error"
           }
         }
+
+  - method: POST
+    path: /abandon
+    response:
+      statusCode: 200
+      content: |
+        {}
```
Comment (Caitlin-cooling) [inline tests/imposter/private-api-config.yaml:199]: I see we have a unit test for the failure scenario for this API call, do we want a browser test too?

File: tests/unit/src/app/check/controllers/abandon.test.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
+const BaseController = require("hmpo-form-wizard").Controller;
+const Controller = require("../../../../../../src/app/check/controllers/abandon");
+
+const {
+  API: {
+    PATHS: { ABANDON },
+  },
+} = require("../../../../../../src/lib/config");
+
+jest.mock();
```
Comment (Caitlin-cooling) [inline tests/unit/src/app/check/controllers/abandon.test.js:10]: Not sure if I'm missing something here but what are we mocking? I've never seen `jest.mock()` called like this before

File: src/locales/en/pages.yml [hunk 5]
```diff
@@ -6,3 +6,6 @@ could-not-match-national-insurance:
   content:
     - We could not find a match for your National Insurance number when we checked with HMRC.
     - You can check your National Insurance number and try again. You’ll only be able to do this once.
+abandon:
+  content: " "
+  h1: "What would you like to do?"
```
Comment (Caitlin-cooling) [inline src/locales/en/pages.yml:11]: The page title seems to be an untranslated string, are we maybe missing a content key? Or is that going to be handled elsewhere?

image

File: tests/browser/pages/abandon.js (hunk 2 above)
Comment (MarieseMikely) [reply tests/browser/pages/abandon.js]: updates made

File: tests/unit/src/app/check/controllers/abandon.test.js (hunk 4 above)
Comment (MarieseMikely) [reply tests/unit/src/app/check/controllers/abandon.test.js]: removed this line

File: src/locales/en/pages.yml (hunk 5 above)
Comment (MarieseMikely) [reply src/locales/en/pages.yml]: updated as per last commit

File: src/locales/en/pages.yml [hunk 6]
```diff
@@ -1,8 +1,14 @@
...
+  title: "What is your national insurance number - Welcome to GOV.UK  One Login"
 could-not-match-national-insurance:
+  title: "What would you like to do - Welcome to GOV.UK  One Login"
   h1: "We could  not match your National Insurance number"
   content:
     - We could not find a match for your National Insurance number when we checked with HMRC.
     - You can check your National Insurance number and try again. You’ll only be able to do this once.
+abandon:
+  content: " "
+  h1: " "
```
Comment (Caitlin-cooling) [inline src/locales/en/pages.yml:13]: This means that we have an empty H1 on the page generally which is considered an accessibility issue. There is a way to make the radio filedset legend the H1 using the `isPageHeading` param looking at the [HMPO component](https://github.com/HMPO/hmpo-components/blob/756f130b74d78d23f292c448b50a2d081c14d58f/components/hmpo-radios/macro.njk). Happy to pair on this if it's helpful!

image

File: tests/imposter/private-api-config.yaml (hunk 3 above)
Comment (Caitlin-cooling) [reply tests/imposter/private-api-config.yaml:162]: Resolving as this has now been addressed 🎉

File: tests/imposter/private-api.yaml (hunk 1 above)
Comment (MarieseMikely) [reply tests/imposter/private-api.yaml:24]: PR has been merged in the API repo

File: src/views/check/national-insurance-number.njk [hunk 7]
```diff
@@ -1,9 +1,25 @@
...
-{% block submitButton %}
-  {{ hmpoSubmit(ctx, {id: "continue", text: translate("buttons.next")}) }}
+{% block mainContent %}
+  {% call hmpoForm(ctx) %}
+      {{ hmpoText(ctx, {
+            id: "nationalInsuranceNumber",
+            namePrefix: "nationalInsuranceNumber",
+            label: {
+                isPageHeading: true,
+                classes: "govuk-label--m"
```
Comment (Caitlin-cooling) [inline src/views/check/national-insurance-number.njk:16]: Thanks for making this change to other pages too! I notice this class is different from the others - is that per the design?

File: tests/imposter/private-api-config.yaml [hunk 8]
```diff
@@ -8,47 +8,25 @@ resources:
...
     response:
       statusCode: 201
       template: true
-      content: |
-        {
-          "session_id": "check-error",
-          "state": "sT@t3",
-          "redirect_uri": "http://example.net/return"
-        }
+      content: '{"session_id": "${stores.testStore.cid}","state": "sT@t3","liam": "yes","redirect_uri": "http://example.net/return"}'
```
Comment (Caitlin-cooling) [inline tests/imposter/private-api-config.yaml:29]: Nice refactor! What is the `"liam": "yes"`for?

File: src/views/check/national-insurance-number.njk (hunk 7 above)
Comment (liamStockwellGDS) [reply src/views/check/national-insurance-number.njk:16]: that's because this isn't a radio,

File: tests/imposter/private-api-config.yaml (hunk 8 above)
Comment (liamStockwellGDS) [reply tests/imposter/private-api-config.yaml]: has been removed, was a debug message

**Review by Caitlin-cooling — APPROVED** [review]
All looks good to me!

**Review by merlinc — APPROVED** [review]
LGTM

**Pull request merged**